presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -270,8 +270,8 @@ private void processNewSources()
 
                 // check if operator is blocked
                 Operator current = operators.get(0);
-                ListenableFuture<?> blocked = isBlocked(current);
-                if (!blocked.isDone()) {
+                if (isBlocked(current)) {
+                    ListenableFuture<?> blocked = getBlockedFuture(current);
```

Comment:
I changed my implementation so that what you propose is possible and satisfies both: no extra futures, no race.